ipa-join: implement calls to JSON-RPC endpoints

Carbenium opened the pull request:
Based on work done by @mulatinho in PR #3544

abbra commented:
```
autoheader: warning: missing template: WITH_IPA_JOIN_XML
autoheader: Use AC_DEFINE([WITH_IPA_JOIN_XML], [], [Description])
autoreconf: /usr/bin/autoheader failed with exit status: 1
configure: error: cannot find install-sh, install.sh, or shtool in "." "./.." "./../.."
```

rcritten commented on freeipa.spec.in:
```diff
@@ -540,6 +541,7 @@ Requires: initscripts
 %endif
 Requires: libcurl >= 7.21.7-2
 Requires: xmlrpc-c >= 1.27.4
+Requires: jansson >= 2.12
```
Is there a reason you are requiring 2.12? Does it provide some minimum level for bug fixing or is it just the current release?

Carbenium replied:
It's the current release available in F32 and Debian Buster. The [changelog](https://jansson.readthedocs.io/en/2.13/changes.html) mentioned some bugfixes for `json_pack`, although I didn't take a detailed look.
No reason otherwise.

Carbenium replied:
Constraint removed.

rcritten commented on configure.ac:
```diff
@@ -179,40 +179,10 @@ if test x"$try_xmlrpc_fallback" = xtrue; then
-    AC_SUBST(JSON_LIBS)
-fi
+PKG_CHECK_MODULES([JANSSON], [jansson])
+PKG_CHECK_MODULES([LIBCURL], [libcurl])
```
Are both required since at build time only one will be used?

Carbenium replied:
No. I'm going to add a conditional based on `WITH_IPA_JOIN_XML`. Should this also be reflected in the spec file?

Carbenium replied:
I'm revising my previous comment. xmlrpc is still needed, since the unenroll code is not converted yet.

Carbenium replied:
Unenrollment is now implemented as well and all XML-RPC related code behind ifdefs.

rcritten commented on client/ipa-join.c:
```diff
@@ -612,6 +616,231 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    /* Perform the call and check for errors */
+    res = curl_easy_perform(curl);
+    if (res != CURLE_OK)
+    {
```
This does not react to expired tickets properly. It instead tries to parse the output of the returned login page as json and fails. The output of ipa-join -d contains:

```
> POST /ipa/json HTTP/1.1
Host: ipa.example.test
referer: https://ipa.example.test/ipa
User-Agent: ipa-join/4.9.0.dev202006011924+gita7b425d68
Accept: application/json
Content-Type: application/json
Content-Length: 131

* upload completely sent off: 131 out of 131 bytes
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Date: Mon, 01 Jun 2020 19:46:03 GMT
< Server: Apache/2.4.41 (Fedora) OpenSSL/1.1.1d mod_wsgi/4.6.4 Python/3.7 mod_auth_gssapi/1.6.1
* gss_init_sec_context() failed: Ticket expired. 
< WWW-Authenticate: Negotiate
< X-Frame-Options: DENY
< Content-Security-Policy: frame-ancestors 'none'
< Last-Modified: Fri, 01 May 2020 17:21:53 GMT
< Accept-Ranges: bytes
< Content-Length: 2714
< Content-Type: text/html; charset=UTF-8
```
I think that in addition to the if (res != CURLE_OK) check you'll need to check the HTTP response code.

To test do: 
```
# kinit -l 1 admin
# ipa-join -d
```

Carbenium replied:
Fixed.

tiran commented on client/ipa-join.c:
```diff
@@ -612,8 +617,246 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+
 static int
-unenroll_host(const char *server, const char *hostname, const char *ktname, int quiet)
+join_krb5_jsonrpc(const char *ipaserver, char *hostname, char **hostdn, const char **princ, bool force, bool quiet) {
```
This function is long and uses a lot of variables. Both make it harder to review the function and track down memory leaks. Could you please split the function into smaller helper functions, e.g. init curl, build the request, parse response, ...?

Carbenium replied:
Done. I hope... :wink:

tiran commented on client/ipa-join.c:
```diff
@@ -612,8 +617,246 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    ASPRINTF((char **)hostdn, "%s", tmp_hostdn);
+
+    char* tmp_princ;
+    if (json_unpack_ex(json_object_get(j_props, "krbprincipalname"), &j_error, JSON_STRICT, "[s]", &tmp_princ) != 0) {
```
This is unsafe, ``(json_object_get(j_props, "krbprincipalname")`` can fail and return NULL.

tiran commented on client/ipa-join.c:
```diff
@@ -612,8 +617,246 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    j_result = json_object_get(j_root, "result");
+
+    char* tmp_hostdn;
+    if (json_unpack_ex(j_result, &j_error, JSON_STRICT, "[s, o]", &tmp_hostdn, &j_props) != 0) {
```
AFAIK ``json_unpack_ex`` supports unpacking of complex, nested objects. You should be able to extract hostdn, krbprincipalname, and krblastpwdchange in one call.

Carbenium replied:
Yes indeed. I somehow misunderstood the docs/API. Should be much nicer now.

Carbenium commented on client/ipa-join.c:
```diff
@@ -612,6 +616,231 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    size_t realsize = size * nmemb;
+    struct curl_buffer *p = (struct curl_buffer *) userdata;
+
+    p->payload = (char *) realloc(p->payload, p->size + realsize + 1);
```
Yes indeed. Fixed.

Carbenium commented on client/ipa-join.c:
```diff
@@ -612,6 +616,231 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+        host = strdup(hostname);
+    }
+
+    curl_global_init(CURL_GLOBAL_DEFAULT);
```
Fixed.

Carbenium commented on client/ipa-join.c:
```diff
@@ -612,6 +616,231 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    if (hostname == NULL) {
+        host = strdup(uinfo.nodename);
+    } else {
+        host = strdup(hostname);
```
Fixed.

Carbenium commented on client/ipa-join.c:
```diff
@@ -612,6 +616,231 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    CURL_SETOPT(curl, CURLOPT_URL, url);
+
+    ASPRINTF(&referer, "referer: https://%s/ipa", ipaserver);
+    headers = curl_slist_append(headers, referer);
```
Fixed.

Carbenium commented on client/ipa-join.c:
```diff
@@ -612,6 +616,231 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+}
+
+static int
+join_krb5_jsonrpc(const char *ipaserver, char *hostname, char **hostdn, const char **princ, int force, int quiet) {
```
I hope the current version is easier to digest :wink:

Carbenium commented on client/ipa-join.c:
```diff
@@ -34,6 +35,9 @@
+#include <limits.h>
 
 #include "xmlrpc-c/base.h"
 #include "xmlrpc-c/client.h"
```
Done.

Carbenium commented on client/ipa-join.c:
```diff
@@ -612,8 +617,246 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+        goto cleanup;
+    }
+
+    j_result = json_object_get(j_root, "result");
```
I actually left this one out since the following `json_unpack_ex` can handle `NULL` case as well and print an error message.

Carbenium commented:
@rcritten @tiran Thanks for the reviews so far! It's been a while I had been in contact with C...

I think/hope I addressed all your comments - if you want to have another go.
JSON-RPC based unenrollment is now implemented as well, so xmlrpc is completely optional.

tiran commented on client/ipa-join.c:
```diff
@@ -34,9 +35,14 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <sys/wait.h>
+#include <curl/curl.h>
+#include <jansson.h>
+#include <limits.h>
 
+#ifdef WITH_IPA_JOIN_XML
 #include "xmlrpc-c/base.h"
 #include "xmlrpc-c/client.h"
+#endif
```
```suggestion
#include <limits.h>

#ifdef WITH_IPA_JOIN_XML
#include "xmlrpc-c/base.h"
#include "xmlrpc-c/client.h"
#else
#include <curl/curl.h>
#include <jansson.h>
#endif
```

Carbenium replied:
Ah, yes. Good catch. Fixed.

Carbenium commented:
@tiran Issues found by clang are fixed. I'm going to move the fixes to the respective commits once you reviewed them.

abbra commented on client/ipa-join.c:
```diff
@@ -612,161 +632,427 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    char *json_str = NULL;
 
-    xmlrpc_env_init(&env);
+    response->payload = (char *) malloc(sizeof(*response->payload));
```
This one makes no sense. payload is a pointer to char, sizeof(*response->payload) is 8 bytes on a typical 64-bit system. The buffer we have in the payload is fileld in by the response callback and there we always do reallocation.
Since you are doing `realloc()` in the response callback anyway, the response buffer can be left as it is (because we pass a zeroed buffer here) or explicitly zeroed instead here. I'd suggest to remove any payoad part handling here completely.

Carbenium replied:
True... Initial malloc has been removed.

abbra commented:
```
checking for JANSSON... yes
checking for LIBCURL... no
configure: error: Package requirements (libcurl) were not met:
Package 'libcurl', required by 'virtual:world', not found
Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.
Alternatively, you may set the environment variables LIBCURL_CFLAGS
and LIBCURL_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
error: Bad exit status from /var/tmp/rpm-tmp.wJxFjt (%build)
RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.wJxFjt (%build)
Child return code was: 1
```

abbra commented on client/Makefile.am:
```diff
@@ -86,6 +86,8 @@ ipa_join_LDADD = 		\
 	$(LDAP_LIBS)		\
 	$(SASL_LIBS)		\
 	$(XMLRPC_LIBS)		\
+	$(JANSSON_LIBS)		\
+	$(LIBCURL_LIBS)		\
```
You have defined automake condition for choosing which join method to use but the condition is not used in the makefile that specifies dependencies. It means both sets of libraries will be pulled in. We need to choose one.
I think this part needs to be split into a separate condition and assign either set of libraries to a variable that then gets referenced here.

Carbenium replied:
I was also under the impression that this would be necessary. However when I look at the linker flags for both cases, only the necessary flags are used.

JSON:
```
[...] ../util/.libs/libutil.a [...] -ljansson -lcurl -lpopt
```

XML:
```
[...] ../util/.libs/libutil.a [...] -lxmlrpc -lxmlrpc_client -lxmlrpc_util -lpopt
```

rcritten replied:
When JSON is used then XMLRPC_CFLAGS and XLRPC_LIBS are both empty:
```
XMLRPC_CFLAGS =
XMLRPC_LIBS =
```
And the inverse is true when --with-ipa-join-xml is selected. the JSAN values are empty.

This looks ok to me. It might be slightly confusing looking only at the client Makefile but in the source-level config.log it's pretty clear.

rcritten commented:
set_user_agent() is only used by xmlrpc-c and should be wrapped with a WITH_IPA_JOIN_XML ifdef.

rcritten commented on client/ipa-join.c:
```diff
@@ -612,161 +632,417 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
     return rval;
 }
 
+#else
```
I assume at some point we're going to completely remove the xmlrpc-c support so it isn't a huge deal but it might be nice to have a comment here that it will use JSON (there is only one #define, XMLRPC or something else).

rcritten commented on client/ipa-join.c:
```diff
@@ -612,161 +632,417 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+
+    /* delegate authentication to GSSAPI */
+    CURL_SETOPT(curl, CURLOPT_GSSAPI_DELEGATION, CURLGSSAPI_DELEGATION_FLAG);
+    CURL_SETOPT(curl, CURLOPT_HTTPAUTH, CURLAUTH_NEGOTIATE);
```
I have the feeling this is delegating the TGT. xmlrpc-c enables it with (simplified):

```
    rc = curl_easy_setopt(curlSessionP, CURLOPT_HTTPAUTH, CURLAUTH_GSSNEGOTIATE)
```

Carbenium replied:
`CURLAUTH_GSSNEGOTIATE` has been deprecated in https://github.com/curl/curl/commit/e38ba430142153e95e2b0abe08a31d626a5c5864 and is an alias for `CURLAUTH_NEGOTIATE`.

rcritten replied:
Ok I have some additional changes but I'll submit them as a separate PR.

rcritten commented:
Can you update the man page to generalize the XML-RPC references? I'm not sure what to suggest, maybe just make it RPC or add a note that the underlying protocol may depend on how your IPA was built or something hand wavy like that.

rcritten commented on client/ipa-join.c:
```diff
@@ -1064,3 +1521,6 @@ main(int argc, const char **argv) {
 }
+
+#undef CURL_SETOPT
+#undef ASPRINTF
```
Just curious, why undefine them?

Carbenium replied:
Just cleaning up after my self ;) Thought it might be a good idea...
But I suppose since it's not a header file, the macro scope is limited to this file. I can remove them if you want me to.

rcritten replied:
Yeah, I don't think they are necessary. Please remove them.

tiran commented on client/ipa-join.c:
```diff
@@ -612,161 +632,417 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    if (curl_easy_setopt(curl, opt, val) != CURLE_OK) { \
+        if (!quiet) \
+            fprintf(stderr, _("curl_easy_setopt() failed\n")); \
+        rval = 17; \
```
What's the meaning of ``17`` here? How about you define global exist code as constants and reuse the named constants?

rcritten replied:
The current join code is filled with numeric return values. 17 is XMLRPC fault in the man page, we could convert it to RPC fault. I'd be ok leaving the numeric values and fixing the constants later as it isn't related to the JSON code.

Carbenium replied:
I would prefer as well to do this separately. I think this PR is already big enough as it is.

rcritten commented:
This is looking good. Triggering full CI to be sure. The Azure failure is unrelated to this PR and is an environmental thing. I'll trigger that too to re-run too.

rcritten commented:
/AzurePipelines run

rcritten commented:
@Carbenium  Can you rebase this against the current master? CI is failing because some code that controls template versions is out-of-date.

Carbenium commented:
@rcritten Sure thing. Here you go.

fcami commented:
/AzurePipelines run

fcami commented:
Triggering Azure CI + PR-CI as the Azure failure is unrelated to the PR.

fcami commented:
/AzurePipelines run

rcritten commented:
All tests are passing, this looks good to me. Last chance @abbra @tiran any final comments before I ack?

abbra commented:
LGTM.

So this is going to be a change in defaults: we would default to using JSON-RPC now unless `configure --with-ipa-join-xml` would be done explicitly.

I think it is worth to document in the release notes (add 'changelog' entry in the ticket') because this client would not be able to enroll to very old servers (IPA v2.x?). No practical change to anything else, though.

abbra commented:
master:

* 6e414d22919822bb6f18b8b03f223de9e08569ef ipa-join: allowing call with jsonrpc into freeipa API
* 5e7e4f0e262caef0bae1518a575bb98f69a9d7e1 ipa-join: don't set TLS related curl options for JSON-RPC
* c197918e8d220f10676f42b3589e47599b7a11e5 ipa-join: improve curl error handling in JSON-RPC code
* c905f94f9b829faec03a016ebd7a344caf3ca144 ipa-join: buffer curl response before parsing json
* 25205f44a10b9e4846121d1a9b05aafa74485292 ipa-join: switch to jansson for json handling
* 677659c8da27de7b256b7067e513f089892d05b8 ipa-join: extract unenrollment code common to JSON and XML-RPC to separate function
* 62503e4fd0e92aae60da49e6917ac62ea52def96 ipa-join: implement JSON-RPC based unenrollment
* f6940772dd9d5f550e9ce4ff20c779864ad4eb68 ipa-join: select {JSON,XML}-RPC at build time
* a1b117a28b670c71e60fdd1c0ec0c6ef8f8f438b ipa-join: Use bool type where appropriate
* 7cc977b993c30add2ba6ec2f823d782cc4d99f7f ipa-join: Generalize XML-RPC references in man page
